File: prj6/Graph.py
class Node:
    def __init__(self, data=None):
        self.data = data
        self.next = None
        self.previous = None


class Graph:

    def __init__(self):
        self.graph_dict = dict()

    def add_vertex(self, vertex):
        self.graph_dict[vertex] = []

    def add_edge(self, firstVertex, secondVertex):
        try:
            key = self.graph_dict[firstVertex]
            key.append(secondVertex)

            key = self.graph_dict[secondVertex]
            key.append(firstVertex)
        except:
            print("Key Not Found !!")

    def remove_edge(self, firstVertex, secondVertex):
        try:
            key = self.graph_dict[firstVertex]
            key.remove(secondVertex)

            key = self.graph_dict[secondVertex]
            key.remove(firstVertex)
        except:
            print("Key Not Found!")

    def remove_vertex(self, vertex):
        del self.graph_dict[vertex]

    def BFS(self, startVertex):

        visited = [False] * (max(self.graph_dict) + 1)

        queue = []

        queue.append(startVertex)
        visited[startVertex] = True

        while queue:

            s = queue.pop(0)
            print(s, end=" ")

            for i in self.graph_dict[s]:
                if visited[i] == False:
                    queue.append(i)
                    visited[i] = True

    def DFS(self, startVertex, visited=None):
        if visited is None:
            visited = []
        visited.append(startVertex)
        vertices = self.graph_dict[startVertex]
        for vertex in vertices:
            if vertex not in visited:
                self.DFS(vertex, visited)
        return visited

    def display_graph(self):
        print(self.graph_dict)

File: prj6/graph.cpp
#include <iostream>
#include <unordered_map>
#include <list>
#include <queue>
#include <stack>
#include <unordered_set>

using namespace std;
class Graph {
private:
    unordered_map<int, list<int>> adjacencyList;

public:
    
    void addVertex(int vertex) {
        if (adjacencyList.find(vertex) == adjacencyList.end()) {
            adjacencyList[vertex] = list<int>();
        }
    }

    
    void addEdge(int vertex1, int vertex2) {
        addVertex(vertex1);
        addVertex(vertex2);

        adjacencyList[vertex1].push_back(vertex2);
        adjacencyList[vertex2].push_back(vertex1);
    }

    void removeEdge(int vertex1, int vertex2) {
        if (adjacencyList.find(vertex1) != adjacencyList.end() &&
            adjacencyList.find(vertex2) != adjacencyList.end()) {
            adjacencyList[vertex1].remove(vertex2);
            adjacencyList[vertex2].remove(vertex1);
        }
    }
    
    void removeVertex(int vertex) {
        if (adjacencyList.find(vertex) != adjacencyList.end()) {
            
            for (int neighbor : adjacencyList[vertex]) {
                adjacencyList[neighbor].remove(vertex);
            }
            adjacencyList.erase(vertex);
        }
    }

    void BFS(int startVertex) {
        unordered_set<int> visited;
        queue<int> q;

        visited.insert(startVertex);
        q.push(startVertex);

        while (!q.empty()) {
            int currentVertex = q.front();
            cout << currentVertex << " ";
            q.pop();

            for (int neighbor : adjacencyList[currentVertex]) {
                if (visited.find(neighbor) == visited.end()) {
                    visited.insert(neighbor);
                    q.push(neighbor);
                }
            }
        }
        cout << "\n";
    }

    void DFS(int startVertex) {
        unordered_set<int> visited;
        DFSUtil(startVertex, visited);
        cout << "\n";
    }

    void DFSUtil(int vertex, unordered_set<int>& visited) {
        visited.insert(vertex);
        cout << vertex << " ";

        for (int neighbor : adjacencyList[vertex]) {
            if (visited.find(neighbor) == visited.end()) {
                DFSUtil(neighbor, visited);
            }
        }
    }

    void printGraph() {
        for (const auto& pair : adjacencyList) {
            cout << "Vertex " << pair.first << " -> ";
            for (int neighbor : pair.second) {
                cout << neighbor << " ";
            }
            cout << "\n";
        }
    }
};